### PR TITLE
Kill AI process of eliminated empire

### PR DESCRIPTION
**Pull request by o01eg**

### CODE REVIEW

File: server/ServerApp.h [hunk 1]
```diff
@@ -314,7 +314,7 @@ class ServerApp : public IApp {
     std::map<int, int>      m_player_empire_ids;                ///< map from player id to empire id that the player controls.
     int                     m_current_turn = INVALID_GAME_TURN; ///< current turn number
     bool                    m_turn_expired = false;             ///< true when turn exceeds its timeout
-    std::vector<Process>    m_ai_client_processes;              ///< AI client child processes
+    std::map<std::string, Process> m_ai_client_processes;              ///< AI client child processes
```
Comment (geoffthemedio) [inline server/ServerApp.h]: maybe add `indexed by player name` here?

File: server/ServerFSM.cpp [hunk 2]
```diff
@@ -3537,8 +3537,8 @@ sc::result ShuttingDownServer::react(const CheckEndConditions& u) {
         DebugLogger(FSM) << "All " << server.m_ai_client_processes.size() << " AIs acknowledged shutdown request.";
 
         // Free the processes so that they can complete their shutdown.
-        for (Process& process : server.m_ai_client_processes)
-        { process.Free(); }
+        for (auto& process : server.m_ai_client_processes)
+        { process.second.Free(); }
```
Comment (geoffthemedio) [inline server/ServerFSM.cpp]: `{}` unnecessary; just indent 4 spaces for a single-liner

File: server/ServerApp.cpp [hunk 3]
```diff
@@ -3707,6 +3707,12 @@ void ServerApp::CheckForEmpireElimination() {
         else if (EmpireEliminated(entry.first, m_universe.Objects())) {
             entry.second->Eliminate();
             RemoveEmpireTurn(entry.first);
+            if (GetEmpireClientType(entry.second->EmpireID()) != Networking::ClientType::CLIENT_TYPE_AI_PLAYER) {
+                auto it = m_ai_client_processes.find(entry.second->PlayerName());
+                if (it != m_ai_client_processes.end()) {
+                    it->second.Kill();
+                }
```
Comment (geoffthemedio) [inline server/ServerApp.cpp:3720]: is it removed from the map somewhere else?

File: server/ServerApp.h (hunk 1 above)
Comment (o01eg) [reply server/ServerApp.h]: Fixed.

File: server/ServerApp.cpp (hunk 3 above)
Comment (o01eg) [reply server/ServerApp.cpp:3720]: Fixed.

Comment (o01eg) [reply server/ServerApp.cpp:3720]: Forgot to push changes.

**Comment by geoffthemedio** [comment]
Should defeating an AI reduce the number ofdisplayed players or running FreeOrionCA processes? It doesn't in this test...

[defeat_ai_test.zip](https://github.com/freeorion/freeorion/files/6114431/defeat_ai_test.zip)

**Comment by o01eg** [comment]
I found why it doesn't kill AI but somehow it got wrong player or empire to kill.

**Comment by geoffthemedio** [comment]
I don't see `ServerApp::CheckForEmpireElimination` in the server log...
[freeoriond.log](https://github.com/freeorion/freeorion/files/6116736/freeoriond.log)

**Comment by o01eg** [comment]
> I don't see `ServerApp::CheckForEmpireElimination` in the server log...
> [freeoriond.log](https://github.com/freeorion/freeorion/files/6116736/freeoriond.log)

```
15:51:07.571419 {0x000007e4} [debug] server : ServerApp.cpp:3710 : ServerApp::CheckForEmpireElimination empire #3 Zone of player AI_2 eliminated
...
15:51:07.597426 {0x000007e4} [debug] FSM : ServerFSM.cpp:305 : ServerFSM::HandleNonLobbyDisconnection : Lost connection to player #2, named "AI_2".
```

Did AI successfully killed? Sometimes it kills wrong AI.

**Comment by o01eg** [comment]
Often I get 
```
18:10:35.254743 {0x00007fcda872a740} [debug] server : ServerApp.cpp:3710 : ServerApp::CheckForEmpireElimination empire #3 Zone of player AI_2 eliminated
...
18:10:35.276994 {0x00007fcda872a740} [debug] FSM : ServerFSM.cpp:305 : ServerFSM::HandleNonLobbyDisconnection : Lost connection to player #2, named "AI_2".
18:10:35.277006 {0x00007fcda872a740} [error] FSM : ServerFSM.cpp:315 : AI Player #2, named "AI_2" quit before empire #2 Binding of player AI_1 was eliminated.
```

**Comment by geoffthemedio** [comment]
> Did AI successfully killed? Sometimes it kills wrong AI.

Not sure what you're asking. Both still appear in the in-game players list.

**Comment by o01eg** [comment]
I used from method to determine AI process name. Fixed it. Now it should kill correct AI.

**Comment by geoffthemedio** [comment]
> I used from method

?

**Comment by o01eg** [comment]
> > I used from method
> 
> ?

 I used wrong method ...

**Comment by geoffthemedio** [comment]
The server logs now indicate that a player is being eliminated
````
17:07:14.587702 {0x00007020} [debug] server : ServerApp.cpp:3711 : ServerApp::CheckForEmpireElimination empire #3 Zone of player #2 eliminated
17:07:14.587702 {0x00007020} [debug] server : Process.cpp:70 : Process::Kill
17:07:14.587702 {0x00007020} [debug] server : Process.cpp:72 : Process::Kill calling m_impl->Kill()
17:07:14.587702 {0x00007020} [debug] server : Process.cpp:77 : Process::Kill calling RequestTermination()
[...]
17:07:14.612707 {0x00007020} [debug] FSM : ServerFSM.cpp:3421 : ProcessingTurn time: 0 s
17:07:14.612707 {0x00007020} [debug] FSM : ServerFSM.cpp:305 : ServerFSM::HandleNonLobbyDisconnection : Lost connection to player #2, named "AI_2".
17:07:14.612707 {0x00007020} [info] FSM : ServerFSM.cpp:321 : AI Player #2, named "AI_2" killed after empire #3 Zone of player AI_2 was eliminated.
17:07:14.613708 {0x00007020} [debug] python : <string>:3 : Python current directory set to C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default\python\chat
17:07:14.613708 {0x00007020} [info] python : chat.py:41 : Chat : [[PLAYER_LEFT_GAME,AI_2]] (255, 255, 255, 255)
17:07:14.613708 {0x00007020} [debug] server : ServerApp.cpp:1596 : ServerApp::GetPlayerInfoMap: Compiling PlayerInfo for each player
17:07:14.613708 {0x00007020} [error] network : ServerNetworking.cpp:556 : PlayerConnection::AsyncWriteMessage(): player id = 2 error #10054 "An existing connection was forcibly closed by the remote host"
17:07:14.613708 {0x00007020} [debug] FSM : ServerFSM.cpp:305 : ServerFSM::HandleNonLobbyDisconnection : Lost connection to player #2, named "AI_2".
17:07:14.613708 {0x00007020} [debug] python : <string>:3 : Python current directory set to C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default\python\chat
17:07:14.614708 {0x00007020} [info] python : chat.py:41 : Chat : [[PLAYER_LEFT_GAME,AI_2]] (255, 255, 255, 255)
````

The empire still appears in the in-game list, including after saving and loading, which seems to work.

**Comment by o01eg** [comment]
I'll squash it to backport to my servers.

**Pull request merged**